planner/functional_dependency/fd_graph.go

```diff
@@ -390,3 +420,279 @@ func (s *FDSet) EquivalenceCols() (eqs []*FastIntSet) {
 	}
 	return eqs
 }
+
+// MakeNotNull modify the FD set based the listed column with NOT NULL flags.
+// Most of the case is used in the derived process after predicate evaluation,
+// which can upgrade lax FDs to strict ones.
+func (s *FDSet) MakeNotNull(notNullCols FastIntSet) {
+	for i := 0; i < len(s.fdEdges); i++ {
+		fd := s.fdEdges[i]
+		if fd.strict {
+			continue
+		}
+		// lax can be made strict if all determinant columns are not null.
+		if fd.from.SubsetOf(notNullCols) {
+			// we don't need to clean the old lax one.
```

Comment:
We don't clean it since we will clear it when adding the strict one. I think tell the reader this fact in the comment would help them understand.

Reply:
Yes, exactly